symengine/dense_matrix.cpp

```diff
@@ -364,76 +364,93 @@ void mul_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
 }
 
 // ---------------------------- Joining Operations ---------------------------//
-void row_join(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+void DenseMatrix::row_join(const DenseMatrix &B)
 {
-    SYMENGINE_ASSERT(A.row_ == B.row_
-                     and (A.row_ == C.row_ and C.col_ == A.col_ + B.col_));
+    this->col_insert(B, col_);
+}
 
-    unsigned row = A.row_, col = A.col_;
+void DenseMatrix::col_join(const DenseMatrix &B)
+{
+    this->row_insert(B, row_);
+}
 
-    for (unsigned i = 0; i < row; i++) {
-        for (unsigned j = 0; j < col; j++) {
-            C.m_[i * (col + B.col_) + j] = A.m_[i * col + j];
+void DenseMatrix::row_insert(const DenseMatrix &B, unsigned pos)
+{
+    SYMENGINE_ASSERT(col_ == B.col_ and pos <= row_)
+
+    unsigned row = row_, col = col_;
+    this->resize(row_ + B.row_, col_);
+
+    for (int i = row - 1; i >= numeric_cast<int>(pos); i--) {
+        for (int j = col - 1; j >= 0; j--) {
+            this->m_[(i + B.row_) * col + j] = this->m_[i * col + j];
         }
     }
-    for (unsigned i = 0; i < row; i++) {
-        for (unsigned j = 0; j < B.col_; j++) {
-            C.m_[i * (col + B.col_) + j + col] = B.m_[i * B.col_ + j];
+
+    for (unsigned i = 0; i < B.row_; i++) {
+        for (unsigned j = 0; j < col; j++) {
+            this->m_[(i + pos) * col + j] = B.m_[i * col + j];
         }
     }
 }
 
-void col_join(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
+void DenseMatrix::col_insert(const DenseMatrix &B, unsigned pos)
 {
-    SYMENGINE_ASSERT(A.col_ == B.col_
-                     and (A.col_ == C.col_ and C.row_ == A.row_ + B.row_));
+    SYMENGINE_ASSERT(row_ == B.row_ and pos <= col_)
 
-    unsigned row = A.row_, col = A.col_;
+    unsigned row = row_, col = col_;
+    this->resize(row_, col_ + B.col_);
 
-    for (unsigned i = 0; i < row; i++) {
-        for (unsigned j = 0; j < col; j++) {
-            C.m_[i * col + j] = A.m_[i * col + j];
+    for (int i = row - 1; i >= 0; i--) {
```

Comment:
https://stackoverflow.com/questions/5458204/unsigned-int-reverse-iteration-with-for-loops